mitmproxy/proxy/layers/http/_http2.py

```diff
@@ -407,7 +408,23 @@ def _handle_event(self, event: Event) -> CommandGenerator[None]:
                 yield from self._handle_event(event)
 
     def _handle_event2(self, event: Event) -> CommandGenerator[None]:
-        if isinstance(event, RequestHeaders):
+        if isinstance(event, Wakeup):
+            remaining = self.context.options.http2_ping_keepalive - time.time() + self.last_activity
+            if remaining <= 0.5:
+                remaining = self.context.options.http2_ping_keepalive
+                # PING frames MUST contain 8 octets of opaque data in the payload. 
+                # A sender can include any value it chooses and use those octets in any fashion.
+                self.h2_conn.ping(b"0" * 8)
+                yield Log(f"Send HTTP/2 keep-alive PING to {human.format_address(self.conn.peername)}")
+                yield SendData(self.conn, self.h2_conn.data_to_send())
+            yield RequestWakeup(remaining)
```

Comment:
Stupid question: Shouldn't we update `last_activity` after sending a PING?

Reply:
Right, that would be more clean.